packages/shared/ReactErrorUtils.js

```diff
@@ -89,12 +92,21 @@ const ReactErrorUtils = {
     return ReactErrorUtils._hasCaughtError;
   },
 
+  hasSuppressedError: function() {
+    return ReactErrorUtils._hasSuppressedError;
+  },
+
   clearCaughtError: function() {
     if (ReactErrorUtils._hasCaughtError) {
       const error = ReactErrorUtils._caughtError;
       ReactErrorUtils._caughtError = null;
       ReactErrorUtils._hasCaughtError = false;
       return error;
+    } else if (ReactErrorUtils._hasSuppressedError) {
```

Comment:
We should still count this as caught error. The only way `_hasSuppressedError` should be used is to remember this in `capturedError` so we can avoid logging it later. The rest of the logic should be the same.

Reply:
I did not understand!
In order to clear the `ReactErrorUtils._hasSuppressedError` we need to check the suppress error exist for not.
so count this as a `_caughtError` I need to set `ReactErrorUtils._hasCaughtError` true
```javascript
if (shouldIgnoreError && error != null) {
        this._hasSuppressedError = true;
        this._hasCaughtError = true;
        this._caughtError = error;
      }
```

Reply:
Yes, that sounds good.

Reply:
also one of the test case if failing `resets instance variables before unmounting failed node` test is failing saying `Expected test not to call console.error().`
am i missing something?
```javascript
Expected test not to call console.error().
    
    If the warning is expected, test for it explicitly by:
    1. Using the .toWarnDev() / .toLowPriorityWarnDev() matchers, or...
    2. Mock it out using spyOnDev(console, 'error') or spyOnProd(console, 'error'), and test that the warning occurs.
    
    The above error occurred in the <Foo> component:
        in Foo (at ReactIncrementalErrorLogging-test.internal.js:229)
        in ErrorBoundary (at ReactIncrementalErrorLogging-test.internal.js:228)
    
    React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.
        at CustomConsole.newMethod [as error] (/home/vikram/code/javascript/oss/react/scripts/jest/setupTests.js:70:19)
        at logCapturedError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberErrorLogger.js:68:13)
        at captureError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1017:53)
        at renderRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:841:22)
        at performWorkOnRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1607:24)
        at performWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1517:7)
        at performAsyncWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1495:5)
        at flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:292:5)
        at flushUnitsOfWork.next (<anonymous>)
        at Object.flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:421:129)
      
      at Object.<anonymous> (scripts/jest/setupTests.js:114:15)

```

Reply:
We want to change [this condition](https://github.com/facebook/react/blob/26185759e4f05025b1fdbe5e238a44aefe5669d4/scripts/jest/setupEnvironment.js#L43-L45) to always call `preventDefault()` (without the condition). Then it shouldn't log at all (since all errors will end up suppressed).

After that, a few other tests will fail that test *logging specifically*. We should figure out a way to exempt those.

Reply:
In out test cases we need to call `preventDefault` when any component throws error intentionally,
but I am not able to figure out how to add a error event handler in `Jest` as` window` object is not there.

Reply:
I don't understand. Can you make a more detailed write-up about cases that are currently failing, and what you need help with?

Reply:
For example [test](https://github.com/facebook/react/blob/master/packages/react-test-renderer/src/__tests__/ReactTestRenderer-test.js#L398-L457) `supports error boundaries` in `ReactTestRenderer-test.js` failing. with an error `Expected test not to call console.error().` This is because we need to call `prevent default` so that out handleSuppressError logic gets executed. 

Here is an [example](https://jsfiddle.net/vikramcse/69z2wepo/97544/) and what I am thinking to add event handler.
But we can not do this because `window` object is not available.

Reply:
>But we can not do this because as window object is not available in jsdom.

Not sure what you mean. It *is* available. There are a few tests that run without it because they run without jsdom in the first place, but generally it's there.

Reply:
In the past comments you mentioned (for writing unit test):

Instead we can try changing the `expect().toThrow()` Jest matcher to set up an 'error' handler that does `preventDefault()`. Or figure out some other way to avoid failing all tests that intentionally throw an error.

I am not able to figure out how to implement this, can you give me some clue?

Reply:
You can add a custom matcher. See `toWarnDev` definition for an example of custom matcher. Maybe there is a way to similarly override the built-in `toThrow` matcher and do some custom logic in it?

Reply:
we have added below code in  `setupEnvironment.js` so that, when an error (intentionally) is thrown in our tests it will call `preventDefault` and eventually out `handleSuppress` logic will get called.

```javascript
if (typeof window !== 'undefined') {
  window.addEventListener('error', event => {
    if (event.error != null) {
      event.preventDefault();
    }
  });
```
But at the time of debugging in this `setupEnvironment.js` the `eventListnener` is not added because `window` object is `undefined`.

1. Is this  `setupEnvironment.js` file get used while running tests?
2. Why I am not able to get `window` object?

Reply:
Some tests (only a few) use `@jest-environment node`. Search for it. Those won’t have a `window`, but they also won’t have this problem either (because the problem is specific to jsdom).

All other tests will have `window`.

Remember `setupEnvironment` executes once per test file, not once per project. So it runs in different environments.

Reply:
File `ReactTestRenderer.js` file has the `@jest-environment node` flag.  It means at the time of initialization `setupEnvironment` the `window` object is not available and therefore `eventListnener` is not initialized.

in which below test is failing.

```javascript
it('supports error boundaries', () => {
    const log = [];
    class Angry extends React.Component {
      render() {
        log.push('Angry render');
        throw new Error('Please, do not render me.');
      }
      componentDidMount() {
        log.push('Angry componentDidMount');
      }
      componentWillUnmount() {
        log.push('Angry componentWillUnmount');
      }
    }

    class Boundary extends React.Component {
      constructor(props) {
        super(props);
        this.state = {error: false};
      }
      render() {
        log.push('Boundary render');
        if (!this.state.error) {
          return (
            <div>
              <button onClick={this.onClick}>ClickMe</button>
              <Angry />
            </div>
          );
        } else {
          return <div>Happy Birthday!</div>;
        }
      }
      componentDidMount() {
        log.push('Boundary componentDidMount');
      }
      componentWillUnmount() {
        log.push('Boundary componentWillUnmount');
      }
      onClick() {
        /* do nothing */
      }
      componentDidCatch() {
        this.setState({error: true});
      }
    }

    const renderer = ReactTestRenderer.create(<Boundary />);
    expect(renderer.toJSON()).toEqual({
      type: 'div',
      props: {},
      children: ['Happy Birthday!'],
    });
    expect(log).toEqual([
      'Boundary render',
      'Angry render',
      'Boundary componentDidMount',
      'Boundary render',
    ]);
  });
```

error:
```javascript
Expected test not to call console.error().
    
    If the warning is expected, test for it explicitly by:
    1. Using the .toWarnDev() / .toLowPriorityWarnDev() matchers, or...
    2. Mock it out using spyOnDev(console, 'error') or spyOnProd(console, 'error'), and test that the warning occurs.
    
    The above error occurred in the <Foo> component:
        in Foo (at ReactIncrementalErrorLogging-test.internal.js:229)
        in ErrorBoundary (at ReactIncrementalErrorLogging-test.internal.js:228)
    
    React will try to recreate this component tree from scratch using the error boundary you provided, ErrorBoundary.
        at CustomConsole.newMethod [as error] (/home/vikram/code/javascript/oss/react/scripts/jest/setupTests.js:70:19)
        at logCapturedError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberErrorLogger.js:68:13)
        at captureError (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1017:53)
        at renderRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:841:22)
        at performWorkOnRoot (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1607:24)
        at performWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1517:7)
        at performAsyncWork (/home/vikram/code/javascript/oss/react/packages/react-reconciler/src/ReactFiberScheduler.js:1495:5)
        at flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:292:5)
        at flushUnitsOfWork.next (<anonymous>)
        at Object.flushUnitsOfWork (/home/vikram/code/javascript/oss/react/packages/react-noop-renderer/src/ReactNoop.js:421:129)
      
      at Object.<anonymous> (scripts/jest/setupTests.js:114:15)
```
Expected result would be error should not be printed in the console.
```javascript
function onError(event) {
        error = event.error;
        didSetError = true;
        if (error === null && event.colno === 0 && event.lineno === 0) {
          isCrossOriginError = true;
        }
        if (event.defaultPrevented) {
          shouldIgnoreError = true;
        }
      }
```
I checked the reason it's happening because in `invokeGuardedCallback.js`
the `onError` (above) function is not called so the `shouldIgnoreError` flag remains `false`.
Does this problem trace sounds good?

Reply:
Yes, thanks!

I think for that test (and the few other tests that also have this problem) we can maybe wrap them in `expect(...).toWarnDev()`?

Reply:
By adding `expect(...).toWarnDev()` the tests are now passing  on my machine :)